noxfile.py

```diff
@@ -49,6 +50,43 @@
 nox.options.sessions = ["qa"]
 
 
+def get_gcc_version_to_use():
+    CC = os.environ.get("CC")
+    if CC is None:
+        for gcc in reversed([gcc for gcc in GCC_VERSIONS]):
+            gcc = shutil.which(gcc)
+            if gcc is not None:
+                CC = os.path.split(gcc)[1]
+                break
+        else:
+            CC = shutil.which("gcc")
+            if CC is None:
+                raise RuntimeError(
+                    f"No valid compiler found, searched for gcc, {', '.join(GCC_VERSIONS)}."
+                )
+
+            # gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
+            # Copyright (C) 2019 Free Software Foundation, Inc.
+            # This is free software; see the source for copying conditions.  There is NO
+            # warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+            output = subprocess.check_output([CC, "--version"]).decode().split("\n")
+            if output[0].startswith("gcc"):
+                version = output[0].split(" ")[-1]
+                major_version = version.split(".")[0]
+                CC = f"gcc-{major_version}"
+            elif " clang version " in output[0]:
+                version = output[0].split(" clang version ")[1]
+                major_version = version.split(".")[0]
+                CC = f"clang-{major_version}"
```

Comment:
I think this could be more readable with a regex (famous last words).

```suggestion
            # look for a line "gcc WHATEVER VERSION.WHATEVER"
            search_gcc_version = re.search(r"^gcc\b.* ([0-9]+)\.\S+$", output, re.M)
            
            # look for a line "WHATEVER clang version VERSION.WHATEVER"
            search_clang_version = re.search(r"\bclang version ([0-9]+)\.")
            
            if search_gcc_version:
                major_version = search_gcc_version.group(1)
                CC = f"gcc-{major_version}"
            elif search_clang_version:
                major_version = search_clang_version.group(1)
                CC = f"clang-{major_version}"
```